src/compiler/scanner/lines_1000_2000.rs

```diff
@@ -23,44 +30,188 @@ impl Scanner {
         ret.push_str(&self.text_substring(start, self.pos()));
         ret
     }
 
-    pub(super) fn scan_number(&self) -> ScanNumberReturn {
+    pub(super) fn scan_number(&self, on_error: Option<ErrorCallback>) -> ScanNumberReturn {
         let start = self.pos();
-        let main_fragment = self.scan_number_fragment();
-        let end = self.pos();
-        let result: String = self.text_substring(start, end);
+        let main_fragment = self.scan_number_fragment(on_error);
+        let mut decimal_fragment: Option<String> = None;
+        let mut scientific_fragment: Option<String> = None;
+        if self.text_char_at_index(self.pos()) == CharacterCodes::dot {
+            self.increment_pos();
+            decimal_fragment = Some(self.scan_number_fragment(on_error));
+        }
+        let mut end = self.pos();
+
+        if matches!(
+            self.maybe_text_char_at_index(self.pos()),
+            Some(CharacterCodes::E | CharacterCodes::e)
+        ) {
+            self.increment_pos();
+            self.add_token_flag(TokenFlags::Scientific);
+            if matches!(
+                self.maybe_text_char_at_index(self.pos()),
+                Some(CharacterCodes::plus | CharacterCodes::minus)
+            ) {
+                self.increment_pos();
+            }
+            let pre_numeric_part = self.pos();
+            let final_fragment = self.scan_number_fragment(on_error);
+            if final_fragment.is_empty() {
+                self.error(on_error, &Diagnostics::Digit_expected, None, None);
+            } else {
+                scientific_fragment = Some(format!(
+                    "{}{}",
+                    self.text_substring(end, pre_numeric_part),
+                    final_fragment
+                ));
+                end = self.pos();
+            }
+        }
+        let mut result: String;
+        if self.token_flags().intersects(TokenFlags::ContainsSeparator) {
+            result = main_fragment;
+            if let Some(decimal_fragment) = decimal_fragment.as_ref() {
+                result.push_str(&format!(".{}", decimal_fragment));
+            }
+            if let Some(scientific_fragment) = scientific_fragment {
+                result.push_str(&scientific_fragment);
+            }
+        } else {
+            result = self.text_substring(start, end);
+        }
 
-        if false {
-            unimplemented!()
+        if decimal_fragment.is_some() || self.token_flags().intersects(TokenFlags::Scientific) {
+            self.check_for_identifier_start_after_numeric_literal(
+                on_error,
+                start,
+                Some(
+                    decimal_fragment.is_none()
+                        && self.token_flags().intersects(TokenFlags::Scientific),
+                ),
+            );
+            ScanNumberReturn {
+                type_: SyntaxKind::NumericLiteral,
+                value: result.parse::<f64>().unwrap().to_string(),
```

Comment:
The Typescript version uses `+` coercion to convert this string to a number, it looks like since Rust's `String.parse::<f64>()` can parse eg scientific notation it should be able to successfully convert from the number-string to a number